### ISSUE TITLE
3D border colors don't match gecko (and are too naive)

### ISSUE THREAD

**Issue by Gankro** [servo/webrender#1280]
See e.g. layout/reftests/w3c-css/received/css-multicol-1/multicol-rule-groove-000-ref.xht, where we render the "light" color darker than gecko.

Gecko's impl can be found in `NS_GetSpecial3DColors`, ours can be found in `ps_border_edge.vs.glsl:write_color()`.

It looks like we agree with chrome (unverified), so this is clearly a fuzzy implementation detail. We should just be sure we're fine with changing firefox's behaviour. (this causes some transient reftest failures when a test case uses gecko but the reference uses webrender)

Example Renderings - left to right: chrome, webrender, gecko:
![screen shot 2017-05-19 at 3 42 02 pm](https://cloud.githubusercontent.com/assets/1136864/26264152/cdd1528a-3ca9-11e7-8fdc-feb77a017032.png)

**Comment by Gankro** [comment]
Ah I just found a much better case, which demonstrates why complex logic exists here (and that we don't match chrome): We render a "black" 3d border as pure black, but everyone else significantly lightens it so that there's still a 3d effect.

See: layout/reftests/table-bordercollapse/border-style-outset-becomes-groove-ref.html

**Comment by Gankro** [comment]
Hillariously, it seems the legacy table property `frame="..."` is lowered to a 3d border so this also breaks all reftests checking for that.

**Comment by kvark** [comment]
TL;DR: 3D borders need to take the background color into account when producing the color

One way to get it would be to explore various blending modes and maybe find an approximation. Given that Blink and Gecko lighten up differently, approximation should be allowed here.

Another way, a more robust one, would be to have the borders spawning a render task for the background. This is similar to the way our box shadows are implemented. With the background in a cache target, we'll be able to directly port the Gecko logic in WR shader, albeit sacrificing a bit of performance for the 3D borders case.

**Comment by glennw** [comment]
I set up a series of border tests with various colors, styles and backgrounds (HTML and outputs below) and compared on Gecko, Blink and Servo.

It looks like Blink doesn't take the background color into account (except for normal alpha blending), and would thus be far simpler to match.

What is our policy here? Do we need to exactly match Gecko on unspecified behavior or would it be reasonable to use a simpler implementation that matches Blink in this case?

Blink:

![border-chrome](https://user-images.githubusercontent.com/2350882/27701548-0a9d4440-5cb6-11e7-8c04-a54b1e53f708.png)

Gecko:

![border-gecko](https://user-images.githubusercontent.com/2350882/27701547-0a961a76-5cb6-11e7-8e3a-947f562dd258.png)

Servo:

![border-servo](https://user-images.githubusercontent.com/2350882/27701546-0a922dda-5cb6-11e7-98d4-00086976d249.png)


```
<!DOCTYPE html>
<html>
  <head>
    <title>3D border colors</title>
    <style type="text/css">
      div > div > div {
        width: 16px;
        height: 16px;
        border-width: 16px;
        display: inline-block;
        margin: 4px;
      }

      .ss {
        border-style: solid;
      }

      .si {
        border-style: inset;
      }

      .so {
        border-style: outset;
      }

      .sg {
        border-style: groove;
      }

      .sr {
        border-style: ridge;
      }

      .c0 {
        border-color: rgba(255, 0, 0, 1);
      }

      .c1 {
        border-color: rgba(0, 255, 0, 1);
      }

      .c2 {
        border-color: rgba(0, 0, 255, 1);
      }

      .c3 {
        border-color: rgba(255, 255, 0, 1);
      }

      .c4 {
        border-color: rgba(0, 0, 0, 1);
      }

      .c5 {
        border-color: rgba(255, 255, 255, 1);
      }

      .c6 {
        border-color: rgba(128, 128, 128, 1);
      }

      .c7 {
        border-color: rgba(255, 0, 0, 0.4);
      }

      .c8 {
        border-color: rgba(0, 255, 0, 0.4);
      }

      .c9 {
        border-color: rgba(0, 0, 255, 0.4);
      }

      .c10 {
        border-color: rgba(255, 255, 0, 0.4);
      }

      .c11 {
        border-color: rgba(0, 0, 0, 0.4);
      }

      .c12 {
        border-color: rgba(255, 255, 255, 0.4);
      }

      .c13 {
        border-color: rgba(128, 128, 128, 0.4);
      }
    </style>
  </head>
  <body>
    <div style="background: white">
      <div>
        <div class="ss c0"></div>
        <div class="ss c1"></div>
        <div class="ss c2"></div>
        <div class="ss c3"></div>
        <div class="ss c4"></div>
        <div class="ss c5"></div>
        <div class="ss c6"></div>
        <div class="ss c7"></div>
        <div class="ss c8"></div>
        <div class="ss c9"></div>
        <div class="ss c10"></div>
        <div class="ss c11"></div>
        <div class="ss c12"></div>
        <div class="ss c13"></div>
      </div>
      <div>
        <div class="si c0"></div>
        <div class="si c1"></div>
        <div class="si c2"></div>
        <div class="si c3"></div>
        <div class="si c4"></div>
        <div class="si c5"></div>
        <div class="si c6"></div>
        <div class="si c7"></div>
        <div class="si c8"></div>
        <div class="si c9"></div>
        <div class="si c10"></div>
        <div class="si c11"></div>
        <div class="si c12"></div>
        <div class="si c13"></div>
      </div>
      <div>
        <div class="so c0"></div>
        <div class="so c1"></div>
        <div class="so c2"></div>
        <div class="so c3"></div>
        <div class="so c4"></div>
        <div class="so c5"></div>
        <div class="so c6"></div>
        <div class="so c7"></div>
        <div class="so c8"></div>
        <div class="so c9"></div>
        <div class="so c10"></div>
        <div class="so c11"></div>
        <div class="so c12"></div>
        <div class="so c13"></div>
      </div>
      <div>
        <div class="sr c0"></div>
        <div class="sr c1"></div>
        <div class="sr c2"></div>
        <div class="sr c3"></div>
        <div class="sr c4"></div>
        <div class="sr c5"></div>
        <div class="sr c6"></div>
        <div class="sr c7"></div>
        <div class="sr c8"></div>
        <div class="sr c9"></div>
        <div class="sr c10"></div>
        <div class="sr c11"></div>
        <div class="sr c12"></div>
        <div class="sr c13"></div>
      </div>
      <div>
        <div class="sg c0"></div>
        <div class="sg c1"></div>
        <div class="sg c2"></div>
        <div class="sg c3"></div>
        <div class="sg c4"></div>
        <div class="sg c5"></div>
        <div class="sg c6"></div>
        <div class="sg c7"></div>
        <div class="sg c8"></div>
        <div class="sg c9"></div>
        <div class="sg c10"></div>
        <div class="sg c11"></div>
        <div class="sg c12"></div>
        <div class="sg c13"></div>
      </div>
    </div>

    <div style="background: black">
      <div>
        <div class="ss c0"></div>
        <div class="ss c1"></div>
        <div class="ss c2"></div>
        <div class="ss c3"></div>
        <div class="ss c4"></div>
        <div class="ss c5"></div>
        <div class="ss c6"></div>
        <div class="ss c7"></div>
        <div class="ss c8"></div>
        <div class="ss c9"></div>
        <div class="ss c10"></div>
        <div class="ss c11"></div>
        <div class="ss c12"></div>
        <div class="ss c13"></div>
      </div>
      <div>
        <div class="si c0"></div>
        <div class="si c1"></div>
        <div class="si c2"></div>
        <div class="si c3"></div>
        <div class="si c4"></div>
        <div class="si c5"></div>
        <div class="si c6"></div>
        <div class="si c7"></div>
        <div class="si c8"></div>
        <div class="si c9"></div>
        <div class="si c10"></div>
        <div class="si c11"></div>
        <div class="si c12"></div>
        <div class="si c13"></div>
      </div>
      <div>
        <div class="so c0"></div>
        <div class="so c1"></div>
        <div class="so c2"></div>
        <div class="so c3"></div>
        <div class="so c4"></div>
        <div class="so c5"></div>
        <div class="so c6"></div>
        <div class="so c7"></div>
        <div class="so c8"></div>
        <div class="so c9"></div>
        <div class="so c10"></div>
        <div class="so c11"></div>
        <div class="so c12"></div>
        <div class="so c13"></div>
      </div>
      <div>
        <div class="sr c0"></div>
        <div class="sr c1"></div>
        <div class="sr c2"></div>
        <div class="sr c3"></div>
        <div class="sr c4"></div>
        <div class="sr c5"></div>
        <div class="sr c6"></div>
        <div class="sr c7"></div>
        <div class="sr c8"></div>
        <div class="sr c9"></div>
        <div class="sr c10"></div>
        <div class="sr c11"></div>
        <div class="sr c12"></div>
        <div class="sr c13"></div>
      </div>
      <div>
        <div class="sg c0"></div>
        <div class="sg c1"></div>
        <div class="sg c2"></div>
        <div class="sg c3"></div>
        <div class="sg c4"></div>
        <div class="sg c5"></div>
        <div class="sg c6"></div>
        <div class="sg c7"></div>
        <div class="sg c8"></div>
        <div class="sg c9"></div>
        <div class="sg c10"></div>
        <div class="sg c11"></div>
        <div class="sg c12"></div>
        <div class="sg c13"></div>
      </div>
    </div>

    <div style="background: rgb(100, 100, 100)">
      <div>
        <div class="ss c0"></div>
        <div class="ss c1"></div>
        <div class="ss c2"></div>
        <div class="ss c3"></div>
        <div class="ss c4"></div>
        <div class="ss c5"></div>
        <div class="ss c6"></div>
        <div class="ss c7"></div>
        <div class="ss c8"></div>
        <div class="ss c9"></div>
        <div class="ss c10"></div>
        <div class="ss c11"></div>
        <div class="ss c12"></div>
        <div class="ss c13"></div>
      </div>
      <div>
        <div class="si c0"></div>
        <div class="si c1"></div>
        <div class="si c2"></div>
        <div class="si c3"></div>
        <div class="si c4"></div>
        <div class="si c5"></div>
        <div class="si c6"></div>
        <div class="si c7"></div>
        <div class="si c8"></div>
        <div class="si c9"></div>
        <div class="si c10"></div>
        <div class="si c11"></div>
        <div class="si c12"></div>
        <div class="si c13"></div>
      </div>
      <div>
        <div class="so c0"></div>
        <div class="so c1"></div>
        <div class="so c2"></div>
        <div class="so c3"></div>
        <div class="so c4"></div>
        <div class="so c5"></div>
        <div class="so c6"></div>
        <div class="so c7"></div>
        <div class="so c8"></div>
        <div class="so c9"></div>
        <div class="so c10"></div>
        <div class="so c11"></div>
        <div class="so c12"></div>
        <div class="so c13"></div>
      </div>
      <div>
        <div class="sr c0"></div>
        <div class="sr c1"></div>
        <div class="sr c2"></div>
        <div class="sr c3"></div>
        <div class="sr c4"></div>
        <div class="sr c5"></div>
        <div class="sr c6"></div>
        <div class="sr c7"></div>
        <div class="sr c8"></div>
        <div class="sr c9"></div>
        <div class="sr c10"></div>
        <div class="sr c11"></div>
        <div class="sr c12"></div>
        <div class="sr c13"></div>
      </div>
      <div>
        <div class="sg c0"></div>
        <div class="sg c1"></div>
        <div class="sg c2"></div>
        <div class="sg c3"></div>
        <div class="sg c4"></div>
        <div class="sg c5"></div>
        <div class="sg c6"></div>
        <div class="sg c7"></div>
        <div class="sg c8"></div>
        <div class="sg c9"></div>
        <div class="sg c10"></div>
        <div class="sg c11"></div>
        <div class="sg c12"></div>
        <div class="sg c13"></div>
      </div>
    </div>
  </body>
</html>
```

**Comment by jrmuizel** [comment]
I don't feel strongly about the coloring. However, it might be easier just to match Gecko because of all the tests... @dbaron do you know (or know who would know) if the colors that we choose matter?

**Comment by dbaron** [comment]
Changing the behavior to match blink may well be reasonable (though I'd want to consider it a tad more deeply, perhaps next week), but if we think matching blink is the right thing, we should probably also change the existing Gecko code to do so so that:
 (a) we separate the behavior change from the webrender landing
 (b) we get feedback on any risks of that behavior change sooner rather than later.

(From a code perspective, getting rid of the code that considers the background color would definitely be nice.)

**Comment by Gankro** [comment]
To clarify the matter on tests: there's a few reftests that hardcode border colors, and there's some reftests in our transient state where gecko borders get compared to webrender borders.

**Comment by glennw** [comment]
@dbaron That basically matches what @jrmuizel and I just discussed. We can hold off on this until the Gecko team decides whether to change to match the Blink behaviour. Thanks!

**Comment by dbaron** [comment]
It would help if you did a bit more to drive that change.  (Ideally, drive the whole thing to a conclusion, but at minimum, get a bug filed with some information.)

**Comment by glennw** [comment]
Will do, thanks.

**Comment by glennw** [comment]
Submitted https://bugzilla.mozilla.org/show_bug.cgi?id=1382896 to start the discussion in Bugzilla.

**Comment by gw3583** [comment]
This was fixed in Gecko (https://bugzilla.mozilla.org/show_bug.cgi?id=1382896).
